checker/src/org/checkerframework/checker/lock/wholefiles/LockVisitor.java

```diff
@@ -0,0 +1,580 @@
+        }
+
+        return super.checkOverride(overriderTree, enclosingType, overridden, overriddenType, p) && isValid;
+    }
+
+    // Check the access of the expression of an ArrayAccessTree or
+    // a MemberSelectTree, both of which happen to implement ExpressionTree.
+    // The 'Expression' in checkAccessOfExpression is not the same as that in
+    // 'Expression'Tree - the naming is a coincidence.
+    protected void checkAccessOfExpression(ExpressionTree tree) {
+        boolean isMethodCall = false;
+
+        Kind treeKind = tree.getKind();
+        assert(treeKind == Kind.ARRAY_ACCESS ||
+               treeKind == Kind.MEMBER_SELECT ||
+               treeKind == Kind.IDENTIFIER);
+
+        if (treeKind == Kind.MEMBER_SELECT) {
+            Element treeElement = TreeUtils.elementFromUse(tree);
+
+            if (treeElement != null && treeElement.getKind() == ElementKind.METHOD) { // Method calls are not dereferences.
+                isMethodCall = true;
+            }
+        }
+
+        if (!isMethodCall) {
+            ExpressionTree expr = null;
+
+            switch(treeKind) {
+                case ARRAY_ACCESS:
+                    expr = ((ArrayAccessTree) tree).getExpression();
+                    break;
+                case MEMBER_SELECT:
+                    expr = ((MemberSelectTree) tree).getExpression();
+                    break;
+                default:
+                    expr = tree;
+                    break;
+            }
+
+            Element invokedElement = TreeUtils.elementFromUse(expr);
+
+            AnnotatedTypeMirror receiverAtm = atypeFactory.getReceiverType(tree);
+            Node node = atypeFactory.getNodeForTree(tree);
+
+            // Is the receiver of the expression being accessed the same as the receiver of the enclosing method?
+            boolean receiverIsThatOfEnclosingMethod = false;
+
+            if (node instanceof FieldAccessNode) {
+                Node receiverNode = ((FieldAccessNode) node).getReceiver();
+                if (receiverNode instanceof ExplicitThisLiteralNode ||
+                    receiverNode instanceof ImplicitThisLiteralNode ||
+                    receiverNode instanceof ThisLiteralNode) {
+                    receiverIsThatOfEnclosingMethod = true;
+                }
+            }
+
+            if (expr != null && invokedElement != null && receiverAtm != null) {
+                boolean skipCheckPreconditions = false;
+
+                AnnotationMirror gb = receiverAtm.getAnnotationInHierarchy(GUARDEDBYINACCESSIBLE);
+                if (gb != null) {
+                    if (AnnotationUtils.areSameByClass( gb, checkerGuardedByClass) ||
+                        AnnotationUtils.areSameByClass( gb, javaxGuardedByClass) ||
+                        AnnotationUtils.areSameByClass( gb, jcipGuardedByClass)) {
+                        // Do nothing.
```

Comment:
Can the code in the "if (skipCheckPreconditions == false)" block be moved here?  I'm having a hard time follow this method with the way it is currently written.  (This suggest would remove the need for the skipCheckPreconditions flag, which is also confusing.)